### ISSUE TITLE
Aerospike client 3 not working

### ISSUE THREAD

**Issue by jameswnl** [autumnjolitz/aerospike#1]
Hi, I have installed v3 C client and it seems not supported (failed at get_client() with KeyError(3,0)).  Any idea? Is this still active?
thanks,

**Comment by autumnjolitz** [comment]
Hi,

The implementation for Aerospike 3 is nonexistent, as per the status matrix. Aerospike 2 LibEvent is known to work in production (my employer uses it), but I designed this to be easy to extend (see https://github.com/benjolitz/aerospike/blob/master/aerospike/implementations_as2libevent.py)

Still, since we've just moved our servers to AS3, I plan to get a rudimentary nonblocking interface using pthreads and lock-free queues. On a side note, I contacted Aerospike and learned that Aerospike 3 servers are backwards compatible with Aerospike 2 clients.

In short,

-- Pull requests are more than welcome.

-- As a stopgap, if you need basic Aerospike functionality (version 3 is really defined by aggregates, UDFs), I suggest getting the Aerospike 2 LibEvent 2 C client and trying this library.

**Comment by autumnjolitz** [comment]
And yes, it is still active. I've been tied up in work of late.

**Comment by jameswnl** [comment]
Hi @benjolitz, 

Thanks for responding. Ok, I'll try with ver. 2.  Do you know where I can download ver. 2 C client? There web site only has ver. 3 available.

And thanks for sharing!
James

**Comment by autumnjolitz** [comment]
http://www.aerospike.com/download/client/libevent/2.1.41/

Ben

**Comment by jameswnl** [comment]
Oh you are fast!
somehow overlooked that. Thanks again!

**Comment by jameswnl** [comment]
Hi @benjolitz,

I built the v.2 client.  It default to static link lib and so I change it to use the Makefile.native which produce the .so.

Do you remember you had to do a lot of tweaking to the makefile in order to get it to work?  So far I have to add "-lssl, -levent-2.0" to the LDFLAGS in order to produce a .so which will work with CDLL("libev2citrusleaf-2.0.so.1").  However, the ctypes.util.find_library("ev2citrusleaf-2.0") still doesn't return.

The "ld -lev2citrusleaf-2.0" failed at ld: warning: cannot find entry symbol _start; not setting start address

thanks

**Comment by autumnjolitz** [comment]
Here is the patch I use to build Aerospike LibEvent2 on Linux:

https://gist.github.com/benjolitz/bbdb48ed33eb50711f23

Drop it into the top level source directory for the C driver (citrusleaf-XXX) and apply it with `patch -p1 < patch_name.patch` where patch_name.patch is the content of the above link.

With regards to being recognized by ctypes.util.find_library, placing the shared libraries into /usr/lib usually does the trick.

The weakest part of this entire system is, unfortunately, Aerospike's poor configuration paired with the Python Standard Library's not-so-good ability to locate shared libraries.

Let me know how it goes!

**Comment by jameswnl** [comment]
Hey @benjolitz,

Thanks a lot! Your patch is a life saver!  Though I have several questions, now I can do basic put/get key. 
1. What is the expected signature of callback for put_key()?
2. Somehow I can't get the put_key() to store string value right. 

> > > > client.put_key(cb, "test", "demo", "As", name="AS Inc", address="CA US", id=1001)
> > > > (None, {}, 1, 4294967295) {}
> > > > 
> > > > client.get_key(cb, "test", "demo", "As", 1000)
> > > > (None, {'name': "\xe0 \x94\xbf'\x7f", 'address': "\xe0 \x94\xbf'"}, 1, 2591989) {}

(Where the "cb" callback simply print _args, *_args)
So, the "id" bin is missing and the string bins (name/address) returned as hex code.

However, using the C client cli would retrieve all 3 bins (id is correct, name/address is the same hex code)

Any suggestions?

Also, it seems the batch operations (get_many/put_many) are not done yet, right?

Thanks again for your help.

James

**Comment by autumnjolitz** [comment]
You're welcome. 

The callbacks have the signature of (error_code, bins, generation_val, expiration_val), where error_code is an integer or None, bins are key-value pairs if the operation should return them (via a get), generation and expiration vals are relevant to the generation count and expiration (if specified in seconds).

With regards to your example, delete the key and try placing only string values in it:

Example (in the CLI):

```
$ aerospike-cli --host xxx.xxx.xxx.xxx
/snip
>>> bl_put_key('test', 'demo', "As", name="AS Inc", address="CA US", id='1001')
(None, {}, 4, 4294967295)
>>> bl_get_key('test', 'demo', 'As')
(None, {'id': '1001', 'name': 'AS Inc', 'address': 'CA US'}, 4, 4294967295)
```

I believe you've found a long standing bug (one that I think I understand well enough to tackle now), but let's verify that it is an error inside the integer encoding function by using only string values.

And Batch operations are unimplemented. currently. One consequence of building this driver for production at my employer was that we were forced to create a minimum spanning set of functionality.

I'd like to resolve that by gradually getting missing/buggy functions working perfectly.

**Comment by jameswnl** [comment]
How did you start this "aerospike-cli"? I can't locate it.  So I'm still using the python repl to do that.

And somehow, the string data are not being pulled out correctly as shown below,

> > > > client.put_key(cb, "test", "demo", "As3", name="AS Inc", address="CA US", id="1001")
> > > > (None, {}, 1, 4294967295) {}
> > > > 
> > > > client.get_key(cb, "test", "demo", "As3")
> > > > (None, {'id': '\xe0 \x94\xbf', 'name': "\xe0 \x94\xbf'\x7f", 'address': "\xe0 \x94\xbf'"}, 1, 2591986) {}

**Comment by jameswnl** [comment]
So I think the integer type is being saved correctly as the following test suggests (note my use of different clients to perform put/get),

Using pypy repl to put_key works,

> > > > client.put_key(cb, "test", "demo", "As5", age=999)
> > > > (None, {}, 1, 4294967295) {}

In fact, the integer type data can be retrieved correctly using the out-of-box cli (/opt/aerospike/bin/cli) installed with the server.

[23:50:14 /etc/aerospike]$ cli -n test -o get -s demo -k As5
{'age': 999}

**Comment by autumnjolitz** [comment]
Yes, put_key preserves values perfectly. It's get_key that is broken.

I've found a bug and fixed it in the integer handling code. So get_key with ints as values should work as intended. See cd14dd0e15ec860713e004a26d9896f4933cf1cc

To be less confusing, I recommend the following lines be placed after import:

```
from aerospike.logger import logger
import logging

logger.addHandler(logging.StreamHandler())
logger.handlers.pop(0)
logger.handlers[0].setLevel(logging.DEBUG)
logger.setLevel(logging.DEBUG)
```

That should give you better debugging output.

Ah, you're using PyPy. That might explain the the string issues - I wondered when I'd get bitten by that. Let me look into that.

With the recent bugfix, can you try get_key on the key you placed 'age' in?

`pip uninstall -y aerospike; pip install git+https://github.com/benjolitz/aerospike.git`

**Comment by autumnjolitz** [comment]
Can you give me the output of when you start up `pypy`?

**Comment by autumnjolitz** [comment]
Also, aerospike-cli should be installed within your path.

If you're using a virtualenv, you'll find it in the bin/ directory of your virtualenv.

To manually invoke it, try `python -c 'import tools; tools.setup_cli()' --host xxx.xxx.xxx.xxx` where the xxx's represent the aerospike node.

**Comment by autumnjolitz** [comment]
I've done some poking around. The bug appears on PyPy alright. In some cases, it uses the wrong char\* pointer. That is very wrong. CPython appears to not provoke the error. 

PyPy will place the wrong values for the bins into the Aerospike store. get_key is unaffected. Tracking down.

**Comment by autumnjolitz** [comment]
Ok, for values inside bins, we're using the `ev2citrusleaf_object_dup_str` function as a stopgap. I need to investigate PyPy as I'd rather pin a string in memory, but I need to figure out the most efficient way to do that.

Currently, we'll copy the string into the C heap and avoid the chance of PyPy playing string optimization games.

Fix in 2550b53a0e5b964714d5b117948da6b34fc9c72a

I believe things should work for you now, as we've fixed the integer handling bug for get_key and the string encoding bug for PyPy.

**Comment by jameswnl** [comment]
Here's the pypy output if it is still needed. I'll test the new fix now.
(venvpypy)[16:27:39 /home/jameswong]# pypy
Python 2.7.6 (32f35069a16d819b58c1b6efb17c44e3e53397b2, Aug 19 2014, 23:32:55)
[PyPy 2.3.1 with GCC 4.4.7 20120313 (Red Hat 4.4.7-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.

**Comment by jameswnl** [comment]
Hey Ben, it's working perfectly now!
So from your experience, how is the performance of this async mode query? Do you use an event to notify the main thread to continue or your application is architected async way already?

**Comment by autumnjolitz** [comment]
Excellent.

The original use case for this driver was to allow a mature Tornado service to exceed the performance of a Java pixel processor, needing to meet a minimum 1,500 requests per second per core to justify it. We got 1,700 requests per second per core without optimization (avoid excess allocations, strip out dead code, make logging asynchronous).

We'd use something like `result = yield gen.Task(connection.get_key, namespace, keyset, key_name)` inside a Tornado handler decorated with `@gen.engine`and `@asynchronous`

When work is done, the pthread_t that is doing the work in C will callback into Python by securing the GIL and activating the constant callback (_handle_event_callback). Inside of python, we convert the void\* user_data pointer to something we can look into the registered callback dictionary, run it and remove it.

An early prototype of this would wrap every callback with `ffi.callback(...)` but that proved to be expensive.

I'm confident there are further optimizations that can be made. For starters, the work dispatcher can be moved into C. It's only in Python due to a race condition within the Aerospike driver (it doesn't like having work added to a live event loop, only a stopped one. So we simply pulse the event loop via a Python thread. That technique should be moved into C)
